### PR TITLE
Lint shell scripts on CI (SQPIT-279)

### PR DESCRIPTION
**Pull request by supersven**
To test it run `make shellcheck`.

https://wearezeta.atlassian.net/browse/SQPIT-279

## Checklist

 - [X] The **PR Title** explains the impact of the change.
 - [X] The **PR description** provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] If this PR changes development workflow or dependencies, they have been A) automated and B) documented under docs/developer/. All efforts have been taken to minimize development setup breakage or slowdown for co-workers.
 - [x] **changelog.d** contains the following bits of information ([details](https://github.com/wireapp/wire-server/blob/develop/docs/developer/changelog.md)):
   - [x] A file with the changelog entry in one or more suitable sub-sections. The sub-sections are marked by directories inside `changelog.d`.

### CODE REVIEW

**Comment by smatting** [comment]
If you want to run the task in ci just add the "make shellcheck" line to linters.dhall https://github.com/zinfra/cailleach/pull/1033

**Comment by supersven** [comment]
> If you want to run the task in ci just add the "make shellcheck" line to linters.dhall here https://github.com/zinfra/cailleach/pull/1033

Hey @smatting ,

Thanks for the advice. I will do. But, first I'll have to solve/ignore the lint issues and remove git GitHub action. :smile:

File: Makefile [hunk 1]
```diff
@@ -118,6 +118,11 @@ formatf:
 formatc:
 	./tools/ormolu.sh -c
 
+# lint all shell scripts with ShellCheck
+.PHONY: shellckeck
```
[the remark this answers is not in the document]
Comment (supersven) [reply Makefile:122]: Thanks for spotting this! :+1:

**Review by stephen-smith — APPROVED** [review]

**Comment by supersven** [comment]
I think this could use a second pair of eye-balls, because there are (in most cases) no tests for these scripts and bugs may be hard to find. I tried to be conservative about the changes: All non-trivial linter issues are ignored for now.

Pinging: @akshaymankar , @smatting , @jschaul

File: deploy/services-demo/demo.sh [hunk 2]
```diff
@@ -38,62 +40,68 @@ function list_descendants () {
...
     if [[ ! -f ${SCRIPT_DIR}/resources/zauth/privkeys.txt || ! -f ${SCRIPT_DIR}/resources/zauth/pubkeys.txt ]]; then
         echo "Generate private and public keys (used both by brig and nginz)..."
-        mkdir -p ${SCRIPT_DIR}/resources/zauth/
+        mkdir -p "${SCRIPT_DIR}/resources/zauth/"
         TMP_KEYS=$(mktemp "/tmp/demo.keys.XXXXXXXXXXX")
-        run_zauth -m gen-keypair -i 1 > $TMP_KEYS
-        cat $TMP_KEYS | sed -n 's/public: \(.*\)/\1/p' > ${SCRIPT_DIR}/resources/zauth/pubkeys.txt
-        cat $TMP_KEYS | sed -n 's/secret: \(.*\)/\1/p' > ${SCRIPT_DIR}/resources/zauth/privkeys.txt
+        run_zauth -m gen-keypair -i 1 > "$TMP_KEYS"
+        # shellcheck disable=SC2002
```
[the remark this answers is not in the document]
Comment (supersven) [reply deploy/services-demo/demo.sh:72]: I'm don't like it, too. IMHO obfuscates the flow out data, especially in multi-line statements... I'll disable it.

File: hack/bin/set-chart-image-version.sh [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 USAGE="$0 <docker-tag> <chart-name>..."
 docker_tag=${1?$USAGE}
+# shellcheck disable=SC2124
 charts=${@:2}
```
[the remark this answers is not in the document]
Comment (supersven) [reply hack/bin/set-chart-image-version.sh:6]: That was my premise: *If in doubt, ignore (disable the issue)* :smile:

[the remark this answers is not in the document]
Comment (supersven) [reply hack/bin/set-chart-image-version.sh:6]: BTW, have you seen that each issue code is documented. E.g. https://github.com/koalaman/shellcheck/wiki/SC2124

File: hack/bin/integration-spring-cleaning.sh [hunk 4]
```diff
@@ -1,5 +1,6 @@
 #!/usr/bin/env bash
 
+# shellcheck disable=SC2034
 DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )/.."
```
[the remark this answers is not in the document]
Comment (supersven) [reply hack/bin/integration-spring-cleaning.sh]: Done.

File: deploy/services-demo/create_test_team_scim.sh [hunk 5]
```diff
@@ -170,8 +175,10 @@ SCIM_USER_INVITATION_CODE=$(curl --silent --show-error \
                             -XGET "$BRIG_HOST/i/teams/invitation-code?team=$TEAM_UUID&invitation_id=$SCIM_USER_INVITATION_ID" | jq -r .code
                             )
 
+# shellcheck disable=SC2002
 scimUserPassword=$(cat /dev/urandom | env LC_CTYPE=C tr -dc a-zA-Z0-9 | head -c 8)
 
+# shellcheck disable=SC2034
 REGISTER_ACCEPT=$(cat <<EOF
```
[the remark this answers is not in the document]
Comment (supersven) [reply deploy/services-demo/create_test_team_scim.sh]: Done.

File: hack/bin/set-helm-chart-version.sh [hunk 6]
```diff
@@ -24,6 +24,7 @@ function write_versions() {
     # update all dependencies, if any
     if [ -a requirements.yaml ]; then
         sed -e "s/  version: \".*\"/  version: \"$target_version\"/g" requirements.yaml > "$tempfile" && mv "$tempfile" requirements.yaml
+        # shellcheck disable=SC2207
         deps=( $(helm dependency list | grep -v NAME | awk '{print $1}') )
```
[the remark this answers is not in the document]
Comment (supersven) [reply hack/bin/set-helm-chart-version.sh]: Looks like this doesn't do the same. :thinking: 

Before:
```sh
➜  wire-server git:(sventennie/shellcheck) ✗ make charts-integration
./hack/bin/copy-charts.sh wire-server
copied /home/sven/src/wire-server/charts/wire-server (and its local dependencies) to /home/sven/src/wire-server/.local/charts/wire-server
./hack/bin/set-wire-server-image-version.sh sven
./hack/bin/set-helm-chart-version.sh "wire-server" 0.0.42
...
```

After:
```sh
➜  wire-server git:(sventennie/shellcheck) ✗ make charts-integration
./hack/bin/copy-charts.sh wire-server
copied /home/sven/src/wire-server/charts/wire-server (and its local dependencies) to /home/sven/src/wire-server/.local/charts/wire-server
./hack/bin/set-wire-server-image-version.sh sven
./hack/bin/set-helm-chart-version.sh "wire-server" 0.0.42
sed: can't read Chart.yaml: No such file or directory
sed: can't read Chart.yaml: No such file or directory
...
```

File: hack/bin/upload-helm-charts-s3.sh [hunk 7]
```diff
@@ -95,10 +95,11 @@ cd "$TOP_LEVEL_DIR"
 
 # If ./upload-helm-charts-s3.sh is run with a parameter, only synchronize one chart
 if [ -n "$chart_dir" ] && [ -d "$chart_dir" ]; then
-    chart_name=$(basename $chart_dir)
+    chart_name=$(basename "$chart_dir")
     echo "only syncing $chart_name"
     charts=( "$chart_name" )
 else
+    # shellcheck disable=SC2207
     charts=( $(make -s -C "$TOP_LEVEL_DIR" echo-release-charts) )
```
[the remark this answers is not in the document]
Comment (supersven) [reply hack/bin/upload-helm-charts-s3.sh:103]: As pushing helm charts might confuse others, I wrote a short test script to show that the outputs are the same:

```bash
#!/usr/bin/env bash

TOP_LEVEL_DIR=$(pwd)

charts1=( $(make -s -C "$TOP_LEVEL_DIR" echo-release-charts) )
echo "${charts1[*]}"

IFS=" " read -r -a charts2 <<< "$(make -s -C "$TOP_LEVEL_DIR" echo-release-charts)"
echo "${charts2[*]}"
```

Result (executed in `wire-server/`):
```
wire-server redis-ephemeral redis-cluster databases-ephemeral fake-aws fake-aws-s3 fake-aws-sqs aws-ingress fluent-bit kibana backoffice calling-test demo-smtp elasticsearch-curator elasticsearch-external elasticsearch-ephemeral minio-external cassandra-external nginx-ingress-controller nginx-ingress-services reaper wire-server-metrics sftd restund coturn
wire-server redis-ephemeral redis-cluster databases-ephemeral fake-aws fake-aws-s3 fake-aws-sqs aws-ingress fluent-bit kibana backoffice calling-test demo-smtp elasticsearch-curator elasticsearch-external elasticsearch-ephemeral minio-external cassandra-external nginx-ingress-controller nginx-ingress-services reaper wire-server-metrics sftd restund coturn
```

File: hack/bin/set-chart-image-version.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (akshaymankar) [reply hack/bin/set-chart-image-version.sh:6]: Yeah, I looked at all of them while reviewing, I just don't understand how that one applies. Its ok I guess we can leave this for now.

**Comment by supersven** [comment]
@akshaymankar , @stephen-smith , I'm very sorry that I wasted your time! But, after spending hours of hunting a subtle bug in `nginz-disco.sh`, I'm convinced that https://github.com/wireapp/wire-server/pull/2361 is a much saver approach.
I'm fearing that some other subtle quoting bug might appear in a stressful situation... and that could be really painful.

**Pull request closed without merging**